### ISSUE TITLE
Webcam / video object detection

### ISSUE THREAD

**Issue by n3ptuve** [eriklindernoren/PyTorch-YOLOv3#97]
Is there a way to test object detection with a cam feed?

**Comment by ccerhan** [comment]
First, you need to install opencv. Then, you can use [VideoCapture](https://docs.opencv.org/3.0-beta/doc/py_tutorials/py_gui/py_video_display/py_video_display.html) to stream images from the webcam. 

Opencv reads the image as a numpy array in BGR order. You have to convert it to RGB first. Darknet allows the input tensor with the size of (batch_size, channel, height, width), e.g. (1, 3, 416, 416). Also you need to resize the input frame as letterbox format which means resizing the image while preserving the aspect ratio applying padding to the remaining areas. This [link](https://github.com/AlexeyAB/darknet/issues/232#issuecomment-336955485) can be helpful for understanding this issue. Here is an example piece of code you need up to the Darknet forward-feed. For the remaining part, you can take a look at the `detect.py`.
```python
# Convert opencv image to pytorch image
frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
x = torch.from_numpy(frame.transpose(2, 0, 1))
x = x.unsqueeze(0).float()  # x = (1, 3, H, W)

# Apply letterbox resize
_, _, h, w = x.size()
ih, iw = (416, 416)
dim_diff = np.abs(h - w)
pad1, pad2 = int(dim_diff // 2), int(dim_diff - dim_diff // 2)
pad = (pad1, pad2, 0, 0) if w <= h else (0, 0, pad1, pad2)
x = F.pad(x, pad=pad, mode='constant', value=127.5) / 255.0
x = F.upsample(x, size=(ih, iw), mode='bilinear') # x = (1, 3, 416, 416)

# Forward-feed to the Darknet
detections = model.forward(x)
# non-max suppression ... and so on
```

**Comment by chunniunai220ml** [comment]
I have done by the above codes @ccerhan ,but I can not get good speed  when test video, the inference time about 50-60ms in 1080Ti(8G).Is there any speed up operations?

**Comment by azikk** [comment]
> First, you need to install opencv. Then, you can use [VideoCapture](https://docs.opencv.org/3.0-beta/doc/py_tutorials/py_gui/py_video_display/py_video_display.html) to stream images from the webcam.
> 
> Opencv reads the image as a numpy array in BGR order. You have to convert it to RGB first. Darknet allows the input tensor with the size of (batch_size, channel, height, width), e.g. (1, 3, 416, 416). Also you need to resize the input frame as letterbox format which means resizing the image while preserving the aspect ratio applying padding to the remaining areas. This [link](https://github.com/AlexeyAB/darknet/issues/232#issuecomment-336955485) can be helpful for understanding this issue. Here is an example piece of code you need up to the Darknet forward-feed. For the remaining part, you can take a look at the `detect.py`.
> 
> ```python
> # Convert opencv image to pytorch image
> frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
> x = torch.from_numpy(frame.transpose(2, 0, 1))
> x = x.unsqueeze(0).float()  # x = (1, 3, H, W)
> 
> # Apply letterbox resize
> _, _, h, w = x.size()
> ih, iw = (416, 416)
> dim_diff = np.abs(h - w)
> pad1, pad2 = int(dim_diff // 2), int(dim_diff - dim_diff // 2)
> pad = (pad1, pad2, 0, 0) if w <= h else (0, 0, pad1, pad2)
> x = F.pad(x, pad=pad, mode='constant', value=127.5) / 255.0
> x = F.upsample(x, size=(ih, iw), mode='bilinear') # x = (1, 3, 416, 416)
> 
> # Forward-feed to the Darknet
> detections = model.forward(x)
> # non-max suppression ... and so on
> ```

hello I copy your code and use the non-max suppression in detect.py
but the detections out put is **[1, 35]** which is expect to be **[1, 10647, 85]** and repo a bug, I have no idea how does this happen

**Comment by codeseeking** [comment]
> I have done by the above codes @ccerhan ,but I can not get good speed when test video, the inference time about 50-60ms in 1080Ti(8G).Is there any speed up operations?

Have you addressed this issue?

**Comment by pvrohin** [comment]
name 'F' is not defined

**Comment by niehen6174** [comment]
Has anyone solved the problem ？？？  I need your help！！！

**Comment by TzuYuOu** [comment]
> name 'F' is not defined

You can add this line to fix your problem
`
import torch.nn.functional as F
`

**Comment by LContini91** [comment]
Could anyone share their full code of their implementation with a webcam? 
That would be very helpful.

**Comment by Flova** [comment]
I closed this issue due to inactivity. Feel free to reopen for further discussion.
